### DISCUSSION TITLE
Create a time series analysis model with river levels

### DISCUSSION THREAD

**Discussion by Jolabbe** [pastas/pastas#355] [Q&A] [answered]
Hello everyone,
I am newbie with time serie analysis but I would like to use it to work on the dataset that I have in my study area. 

So I took the python codes available in the examples doc "Time Series Analysis with Pastas" and then "adding river levels" because the data I would like to analyze is groundwater levels from a piezometer close to a river. It would like to use Pastas to see how much of the level I can explain with rain, PE and then river variations (and perhaps specify the recharge in this area).
Unfortunately it doesn't work and I don't know why but I have some questions: 

How to manage units in Pastas? Do rainfall, PE, and river/water levels have to be in the same units?
About groundwater levels, can Pastas use "m NGF" or in english "m a.s.l" (meters above sea level)?

I think it should work because when you look on the time series I give in this discussion, you clearly sea the response of the groundwater levels simultaneously with the river level...

What do you think about it?

Thank you for your help!
Best regards,
Jordan. 
![raw_data_mezel_d13](https://user-images.githubusercontent.com/95425124/145068568-bedefdcc-b962-4161-89af-1e5062a17d98.png)
.

**Comment by raoulcollenteur** [comment, marked as answer]
Hi Jordan,

Sounds like a nice case study to try and use Pastas on. I think it should be possible to model the time series, using the stresses you mentioned. When you say "it doesn't work", could you specify what exactly doesn't work? Perhaps provide some code and results.

Pastas in principle doesn't know about units, except the non-linear recharge models (e.g., FlexModel, which requires mm/d for the precipitation and potential evaporation).

Cheers,
Raoul

**Reply by martinvonk** [reply]
To add on Raouls comment:

In the example notebook 09 [calibrating pastas models](https://pastas.readthedocs.io/en/latest/examples/009_calibration_options.ipynb.html) you can find the answers to your questions.

For your river level I would thus suggest normalizing the time series by substracting the mean or minimum river water level.

**Reply by Jolabbe** [reply]
Hi Raoul,
Thank you for answering me so quickly!
What I did for now, as you can see in the image, is read the data and stresses with the python code. When I try to run and solve the stress model, the fit as you can see is very, very bad. Surely I must have missed something. Here are the code and the image within:

![Test_mezel_d13](https://user-images.githubusercontent.com/95425124/145073770-ff85ec9c-ae3c-48ab-83c8-c8f4f4d56ac2.png)

import pandas as pd
import pastas as ps
import matplotlib.pyplot as plt

ps.show_versions()
ps.set_log_level("INFO")

oseries = pd.read_csv("C:/Users/jlabbe/Desktop/pastas_d13mezel/mezel_head_d13.csv", parse_dates=True,
                      squeeze=True, index_col=0)
rain = pd.read_csv("C:/Users/jlabbe/Desktop/pastas_d13mezel/mezel_rain.csv", parse_dates=True, squeeze=True,
                   index_col=0)
evap = pd.read_csv("C:/Users/jlabbe/Desktop/pastas_d13mezel/mezel_evap.csv", parse_dates=True, squeeze=True,
                   index_col=0)
waterlevel = pd.read_csv("C:/Users/jlabbe/Desktop/pastas_d13mezel/mezel_riv_allier.csv", parse_dates=True,
                         squeeze=True, index_col=0)

fig, axes = plt.subplots(4,1, figsize=(10, 5), sharex=True)
oseries.plot(ax=axes[0], x_compat=True, legend=True, marker=".", linestyle=" ")
waterlevel.plot(ax=axes[1], x_compat=True, legend=True)
rain.plot(ax=axes[2], x_compat=True, legend=True)
evap.plot(ax=axes[3], x_compat=True, legend=True)
plt.xlim("2018-01-01", "2020-12-31");

ml = ps.Model(oseries.resample("D").mean().dropna(), name="River")

sm = ps.RechargeModel(rain, evap, rfunc=ps.Exponential, name="recharge")
ml.add_stressmodel(sm)

ml.solve(tmin="2018-01-01", tmax="2020-12-31")
ml.plots.results(figsize=(12, 8));

w = ps.StressModel(waterlevel, rfunc=ps.One,  name="waterlevel",
                   settings="waterlevel")
ml.add_stressmodel(w)
ml.solve(tmin="2018-01-01", tmax="2020-12-31")
axes = ml.plots.results(figsize=(12, 8));
axes[-1].set_xlim(0,10);  # By default, the axes between responses are shared.

Thanks a lot for your help, this tool is very interesting indeed.
Cheers to you too,
Jordan.

**Comment by mbakker7** [comment]
What I would do is first start with only the normalized river level as your
stress.
Then select the ps.Gamma as your response.
And turn the noise model off.
That should give a decent fit to start with.

**Comment by Jolabbe** [comment]
Dear all,
Thank you for your many advice, indeed I only used the river level as the main stress (normalized), turned off the noise model and used ps.Gamma, the result is much better.

![model with river as only stress r=90percent](https://user-images.githubusercontent.com/95425124/145230739-dfb40c22-591c-4514-b97f-b87295f3c0b7.png)

I'll keep working on it.
Have a nice day, best regards
Jordan.

**Comment by mbakker7** [comment]
Looks like a good start, Jordan!

Now add rainfall and evaporation and see if that has a significant effect.
If not, take it out again.

Then try the noise model and see if it works. If not, take it out, but that
also means you can probably not make any statistical inference.

**Comment by mbakker7** [comment]
Only thing that bothers me, now that I look more closely, is that the gain
of the step response is 2, which seems odd. That would mean that a 1 m rise
in river level results in a 2 m rise in groundwater. That can not be right.
Are you sure that you normalized the river level correctly by only
subtracting the mean level? Did you include river levels before 2018?

**Reply by Jolabbe** [reply]
Are you talking about the parameter "waterlevel_A" of 1.97? What does it mean?

I took the mean level of this dataset and substracted it for each line, I have some data in 2017 that I used as a warmup to set the first values of 2018.

**Reply by Jolabbe** [reply]
I did another trial with another dataset:

![Figure 2021-12-09 095538](https://user-images.githubusercontent.com/95425124/145368945-b053890f-1d7b-4f10-88da-cff9e52d834d.png)

Here you have a flood in June with a 3m rise of the groundwater head (more or less) and in the stress graph below, there's also a ~3m rise of the river level.
What do you think about it?

**Reply by mbakker7** [reply]
waterlevel_A is the final response, as you can also see in the step response graph in the lower right-hand corner. It is odd that the step is higher than 1. What you may want to do is use the Polder function as your response function and see what you get. That function is specifically designed for the reaction to surface water changes.
